File: src/engine/adapters/postgres/mutate.ts
import type { MutationPlan, MutationResult, MutationRowOp } from '@shared/domain/mutations';
import { encodePath } from '@shared/domain/tree';
import type { Client } from 'pg';
import type { OpCtx } from '../adapter';
import { AdapterError, assertWritable } from '../errors';
import {
  assertAffectedExactlyOne,
  assertColumnsKnown,
  assertKeyIsPrimaryKey,
  orderedOps,
} from '../sql-mutate';
import * as catalog from './catalog';
import { runCommand, runQuery, type TrackQuery } from './query';
import { quoteIdent } from './read';

// Renders one row-op's statement text. preview() (never executes, D6) inlines an escaped SQL
// literal for each value; mutate() pushes a `$n` placeholder onto `params` instead — the two stay
// textually identical in shape (column list, WHERE construction) by sharing this one builder.
type ValueRenderer = (value: string | null, params: unknown[]) => string;

function literalRenderer(value: string | null): string {
  if (value === null) return 'NULL';
  return `'${value.replace(/'/g, "''")}'`;
}

function paramRenderer(value: string | null, params: unknown[]): string {
  params.push(value);
  return `$${params.length}`;
}

function whereFromKey(
  key: Record<string, string | null>,
  render: ValueRenderer,
  params: unknown[],
): string {
  return Object.entries(key)
    .map(([col, val]) =>
      val === null ? `${quoteIdent(col)} IS NULL` : `${quoteIdent(col)} = ${render(val, params)}`,
    )
    .join(' AND ');
}

function renderRowOp(
  relationSql: string,
  op: MutationRowOp,
  render: ValueRenderer,
  params: unknown[],
): string {
  if (op.kind === 'update') {
    const setSql = Object.entries(op.changes)
      .map(([col, val]) => `${quoteIdent(col)} = ${render(val, params)}`)
      .join(', ');
    return `UPDATE ${relationSql} SET ${setSql} WHERE ${whereFromKey(op.key, render, params)}`;
  }
  if (op.kind === 'delete') {
    return `DELETE FROM ${relationSql} WHERE ${whereFromKey(op.key, render, params)}`;
  }
  const columns = Object.keys(op.values);
  const columnSql = columns.map((c) => quoteIdent(c)).join(', ');
  const valueSql = columns.map((c) => render(op.values[c], params)).join(', ');
  return `INSERT INTO ${relationSql} (${columnSql}) VALUES (${valueSql})`;
}

function resolveTablePath(path: MutationPlan['path']): { schema: string; table: string } {
  const [, schemaSegment, objectSegment] = path.segments;
  if (
    path.segments.length !== 3 ||
    schemaSegment?.kind !== 'schema' ||
    !objectSegment ||
    objectSegment.kind !== 'table'
  ) {
    throw new AdapterError(
      'E_NOT_FOUND',
      `mutate requires a database/schema/table path, got: ${encodePath(path.segments)}`,
    );
  }
  return { schema: schemaSegment.name, table: objectSegment.name };
}

// Synchronous (D6): no catalog lookup, no network — trusts the plan's column names as given.
export function preview(plan: MutationPlan): string[] {
  const { schema, table } = resolveTablePath(plan.path);
  const relationSql = `${quoteIdent(schema)}.${quoteIdent(table)}`;
  return orderedOps(plan.ops).map((op) => renderRowOp(relationSql, op, literalRenderer, []));
}

export async function mutate(
  client: Client,
  ctx: OpCtx,
  track: TrackQuery,
  readOnly: boolean,
  plan: MutationPlan,
): Promise<MutationResult> {
  // §8.12's standard: enforced here, not only greyed out in the UI (P5 D11).
  assertWritable(readOnly);

  const { schema, table } = resolveTablePath(plan.path);
  const relationSql = `${quoteIdent(schema)}.${quoteIdent(table)}`;

  // Fresh in this same op (D7, mirrors resolveProjection's P2 D10 discipline) — never trusts a
  // column name the renderer sent without re-checking it against the catalog right now.
  const execSelect: catalog.QueryExecutor = (sql, params) =>
    runQuery(client, sql, params, ctx, track);
  const target = await catalog.getReadTarget(execSelect, schema, table);

  const qualifiedName = `${target.qualifiedName.schema}.${target.qualifiedName.relation}`;
  for (const op of plan.ops) {
    if (op.kind === 'update') {
      assertColumnsKnown(target.columns, [...Object.keys(op.key), ...Object.keys(op.changes)]);
      assertKeyIsPrimaryKey(target.primaryKey, op.key, qualifiedName);
    } else if (op.kind === 'delete') {
      assertColumnsKnown(target.columns, Object.keys(op.key));
      assertKeyIsPrimaryKey(target.primaryKey, op.key, qualifiedName);
    } else {
      assertColumnsKnown(target.columns, Object.keys(op.values));
    }
  }

  const ordered = orderedOps(plan.ops);
  const compiled = ordered.map((op) => {
    const params: unknown[] = [];
    const sql = renderRowOp(relationSql, op, paramRenderer, params);
    return { op, sql, params };
  });
  // One op-log row, one setCommand call, before anything executes (Adapter rule 3, P5 D9).
  ctx.setCommand(
    ordered.map((op) => renderRowOp(relationSql, op, literalRenderer, [])).join(';\n'),
  );

  const execCommand = (sql: string, params: unknown[]) =>
    runCommand(client, sql, params, ctx, track, { suppressCommand: true });

  await execCommand('BEGIN', []);
  let affectedRows = 0;
  try {
    for (const { op, sql, params } of compiled) {
      const { rowCount } = await execCommand(sql, params);
      assertAffectedExactlyOne(op.kind, rowCount);
      affectedRows += rowCount;
    }
    await execCommand('COMMIT', []);
  } catch (err) {
    await execCommand('ROLLBACK', []).catch(() => {});
    throw err;
  }
  return { affectedRows };
}

File: apps/kira-studio/internal/adapters/postgres/mutate.go
package postgres

import (
	"context"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

// literalRenderer adapts sql-mutate.go's LiteralRenderer to the adapters.ValueRenderer shape
// RenderRowOp expects — preview() never touches params, but the signature still takes one, so
// this closure just ignores it.
func literalRenderer(name string, value *string, _ *[]any) (string, error) {
	return adapters.LiteralRenderer(name, value)
}

// resolveTablePath is mutate.ts's own resolveTablePath — postgres's three-segment
// database/schema/table form, distinct from sql-mutate.go's ResolveDatabaseTablePath (the
// two-segment form clickhouse/mysql-family/sqlite share). The database segment stays
// kind-unconstrained (P113 G2), same as before: Mutate's own root check already ran by the time this
// is called.
func resolveTablePath(path model.NodePath) (schema, table string, err error) {
	segs, err := adapters.RequirePath(path, "mutate", adapters.AnySeg("database"), adapters.Seg("schema"), adapters.Seg("table"))
	if err != nil {
		return "", "", err
	}
	return segs[1].Name, segs[2].Name, nil
}

// preview is mutate.ts's preview — synchronous (D6): no catalog lookup, no network, trusts the
// plan's column names as given.
func preview(plan model.MutationPlan) ([]string, error) {
	schema, table, err := resolveTablePath(plan.Path)
	if err != nil {
		return nil, err
	}
	relationSQL := quoteIdent(schema) + "." + quoteIdent(table)
	return adapters.PreviewSQLMutation(plan, relationSQL, literalRenderer, quoteIdent)
}

// mutate is mutate.ts's mutate.
func mutate(ctx context.Context, conn *trackedConn, op *adapters.OpCtx, track TrackQuery, readOnly bool, plan model.MutationPlan) (model.MutationResult, error) {
	resolve := func() (relationSQL, qualifiedName string, columns []model.ColumnMeta, primaryKey []string, err error) {
		schema, table, err := resolveTablePath(plan.Path)
		if err != nil {
			return "", "", nil, nil, err
		}
		// Fresh in this same op (D7, mirrors resolveProjection's P2 D10 discipline) — never trusts a
		// column name the renderer sent without re-checking it against the catalog right now.
		exec := execFor(conn, op, track)
		target, err := getReadTarget(ctx, exec, schema, table)
		if err != nil {
			return "", "", nil, nil, err
		}
		relationSQL = quoteIdent(schema) + "." + quoteIdent(table)
		qualifiedName = target.QualifiedName.Schema + "." + target.QualifiedName.Relation
		return relationSQL, qualifiedName, target.Columns, target.PrimaryKey, nil
	}

	execCommand := func(sql string, params []any) (int64, error) {
		return runCommand(ctx, conn, sql, params, op, track, CommandOptions{SuppressCommand: true})
	}

	// P2 R2: rollback runs on its own detached, timeout-bounded ctx (RunSQLMutation's own rule) so it
	// always reaches the server even when the caller's own ctx is already cancelled — the ad-hoc
	// `_, _ = execCommand("ROLLBACK", nil)` this replaced ran through runCommand's CheckNotStarted,
	// which refuses outright on an already-cancelled ctx, leaving conn (pinned for this adapter's
	// lifetime) stuck inside a stale, still-open transaction for whatever op runs on it next.
	rollback := func(ctx context.Context) {
		// F2: wait for every RunWithAbortRace goroutine this mutate's own statements spawned to
		// actually finish touching conn before issuing ROLLBACK on it — otherwise an aborted
		// statement's own background goroutine (still running conn.Exec) races this cleanup on the
		// same non-concurrency-safe *pgx.Conn.
		conn.waitInFlight()
		cleanupCtx, cancel := context.WithTimeout(context.WithoutCancel(ctx), endTransactionTimeout)
		defer cancel()
		_, _ = conn.Exec(cleanupCtx, "ROLLBACK")
	}

	// §8.12's standard: enforced by RunRelationalMutate's own AssertWritable, not only greyed out in
	// the UI (P5 D11).
	return adapters.RunRelationalMutate(ctx, op, readOnly, plan, adapters.RelationalMutateDeps{
		Resolve: resolve, Quote: quoteIdent, Placeholder: dollarPlaceholder,
		TypeClassFor: typeClassFor, LiteralRenderer: literalRenderer,
		BeginSQL: "BEGIN", Exec: execCommand, Rollback: rollback,
	})
}
